### ISSUE TITLE
Nullable warning on Debug.Assert'd manually-implemented property

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#34792]
**Version Used**: 
3.1.0-beta1-19172-05+edd2de88fb3e84a097fb30b4070e0f219f624e40

**Repro**:
```C#
#nullable enable
using System.Diagnostics;

public class Program
{
    public static void Main() { }

    public struct MyStruct<T>
    {
        private readonly T[]? _array;

        public MyStruct(T[] array)
        {
            _array = array;
            AutoProp = array;
            Field = array;
        }

        public T[]? ManuallyBacked => _array;
        public T[]? AutoProp { get; }
        public T[]? Field;

        public void Method()
        {
            Debug.Assert(Field != null);
            Debug.Assert(Field.Length > 0);

            Debug.Assert(AutoProp != null);
            Debug.Assert(AutoProp.Length > 0);

            Debug.Assert(ManuallyBacked != null);
            Debug.Assert(ManuallyBacked.Length > 0);
        }
    }
}
```

This unexpectedly produces a warning on `ManuallyBacked.Length`:
"Possible dereference of a null reference."

As expected, `AutoProp.Length` and `Field.Length` do not warn.

cc: @cston, @jaredpar, @dotnet/nullablefc

**Comment by cston** [comment]
Thanks for reporting @stephentoub. This is an instance of https://github.com/dotnet/roslyn/issues/29619.

**Comment by stephentoub** [comment]
Thanks, @cston.  Sorry for the dup; it's hard to find matching issues :)
